packages/web/components/JournalyEditor/Toolbar/Toolbar.tsx

```diff
@@ -29,176 +29,209 @@ import ToggleMarkButton from './ToggleMarkButton'
 import ToolbarButton from './ToolbarButton'
 import InsertImageButton from './InsertImageButton'
 import { options, isTableActive } from '../helpers'
+import SwitchToggle from '@/components/SwitchToggle'
+import { useTranslation } from '@/config/i18n'
+import useGetWindowSize from '@/hooks/useGetWindowSize'
+import useIntersectionObserver from '@/hooks/userIntersectionObserver'
 
 type ToolbarProps = {
   allowInlineImages: boolean
+  shouldPlayTypewriterSounds: boolean
+  onToggleShouldPlayTypewriterSounds: () => void
 }
 
-const Toolbar = ({ allowInlineImages }: ToolbarProps) => {
+const Toolbar = ({
+  allowInlineImages,
+  shouldPlayTypewriterSounds,
+  onToggleShouldPlayTypewriterSounds,
+}: ToolbarProps) => {
+  const { t } = useTranslation('j-editor')
+  // The diff between viewport & visualViewport sizes
+  const [viewportsDiff, setViewportsDiff] = useState(0)
+
   const editor = useSlate()
   const isEditorFocused = useFocused()
-  const toolbarRef = useRef<HTMLDivElement>(null)
-  const [isFixed, setIsFixed] = useState(false)
-  // Save the initial distance that the toolbar is from the top of the page
-  const originalElementOffsetTop = toolbarRef?.current?.offsetTop as number
-  // Save height of toolbar so that when it's fixed and removed from document flow,
-  // the parent takes the same height and surrounding content won't collapse
-  const fixedDistanceFromTop = 20
-  const toolbarHeight = toolbarRef.current?.getBoundingClientRect().height || 56
-  const toolbarClasses = classNames('editor-toolbar-container', { 'is-fixed': isFixed })
   const isTableActivated = isEditorFocused && isTableActive(editor)
   const tableIcon = (
     <ToolbarButton type="table" format="insert-table">
       <FormatTableIcon title="Insert table" titleId="toolbar-insert-table-icon" />
     </ToolbarButton>
   )
-  const handleScroll = () => {
-    if (toolbarRef.current) {
-      if (!isFixed && window.pageYOffset >= toolbarRef.current.offsetTop - fixedDistanceFromTop) {
-        setIsFixed(true)
-        return
-      }
-
-      if (isFixed && window.pageYOffset + fixedDistanceFromTop < originalElementOffsetTop) {
-        setIsFixed(false)
-        return
-      }
-    }
-  }
+
+  const windowSize = useGetWindowSize()
+
+  // This is to accounnt for the nav which appears at the top on mobile and has a height of 72px
+  const toolbarStickyOffset =
+    windowSize.width && windowSize.width < navConstants.mobileBreakpoint ? -92 : -20
+
+  const [toolbarObserverRef, toolbarShouldFloat] = useIntersectionObserver({
+    rootMargin: `${toolbarStickyOffset}px 0px 0px 0px`,
+  })
 
   useEffect(() => {
-    window.addEventListener('scroll', handleScroll)
+    const onVisualViewportChange = () => {
+      setViewportsDiff(visualViewport.offsetTop)
+    }
```

Comment:
I kinda think we should debounce this for performance purposes, maybe with an interval of half or a quarter of a second, for performance purposes. But I don't know how much of an impact that's going to have on the feeling while scrolling. There's already some inevitable latency in there so maybe not too bad?

If you have a chance maybe test it with and without the debounce and see if it makes a big difference. But not a blocker.